README.md
# garden Next.js 13.5: Store | Landing | Dashboard (Template 2023)

**garden v1.0.0 is released! Now it's stable and ready for production! 🚀** _If you have used the project before, please double check your ENV file. NextAuth.js, PostgreSQL and next-intl are now used (instead of Clerk, MySQL and next-international)._

## 01. Follow Us Everywhere

[Discord](https://discord.gg/Pb8uKbwpsJ) | [Twitter/𝕏](https://x.com/blefnk) | [GitHub](https://github.com/blefnk) | [Fiverr](https://fiverr.com/blefnk) | [LinkedIn](https://linkedin.com/in/blefnk) | [Facebook](https://facebook.com/blefnk)

## 02. The Bleverse Framework for the Web

[Jump to **Project Features** Section](#04-project-features--roadmap--checklists) | [Jump to **Environment Variables** Section](#12-environment-variables)

**GitHub Goals Progress:**

- [ ] **100 Stars**: 90% [▲▲▲▲▲▲▲▲▲_]
- [ ] **20 Forks**: 75% [▲▲▲▲▲▲▲___]

**We craft the world's most featured, global, and iconic Next.js starter template!** It's really huge! Step into the future with this ultimate Next.js starter! Beautifully created with ❤️ **Next.js 13 template** by **[@blefnk Nazarii Korniienko](https://github.com/blefnk)**, **[Bleverse](https://github.com/bs-oss)**, and its the most gorgeous, entire OSS community. The starter is perfectly opinionated and has i18n, means internationalization, and built with `app router` and `server actions` using `TypeScript` programming language. Everything has been set up for you. It's more than just code; it's a journey. You're not on this journey alone. Download it and move really faster than ever, than anyone!

When you give us the star [on this repo](https://github.com/blefnk/garden) and introduce yourself in the [Discord Channel of the project](https://discord.gg/Pb8uKbwpsJ), you take part in a lotteries where the 5 randomly selected persons will win the early-access to some secret garden features! ⭐ The first gift contest will be started once we reach 100 stars and have at least 5 users participate in the "garden 100 Stars" thread (_just say hello or tell something about yourself there_). This thread can be found in the `bleverse-events` channel [on our Discord server](https://discord.gg/Pb8uKbwpsJ). ✨ _More stars lead to more features!_

**🔥 We are growing so fast! Many thanks [to everyone for the stars](https://github.com/blefnk/garden/stargazers)! Here are the stargazers over time:**

[![Star History Chart](https://api.star-history.com/svg?repos=blefnk/garden&type=Timeline)](https://star-history.com/#blefnk/garden&Timeline)

We've laid the groundwork; now, grab it, dive in and accelerate your development. This is the space where we're playing with everything new in Next.js 13. Stumbled upon something? Use repository discussions, raise an issue right here, or DM us on 𝕏/Twitter [@blefnk](https://x.com/blefnk). Btw, project takes some inspirations from T3 and similar stacks. Our own stack beside other includes also many more other things. Just get it and enjoy! And remember! We love your feedback. We love your stars. So smash it, please! And make your forks! It helps the project to be even coolest! If you've got some coding magic up your sleeve, contributions are always celebrated! 🎉

[📑 Article on Medium about this Project (updates soon)](https://cutt.ly/awf6fScS)

[![Join the Bleverse Discord](https://discordapp.com/api/guilds/1075533942096150598/widget.png?style=banner2)][bleverse-discord]

![Bleverse garden Landing Page Screenshot](/public/screenshot.png)

> **Note**
> This README tries to be really useful. By this reason it requires many of text. It means the some of text may be outdated, and will be updated as we scaling. So please let us know on the [discussion page](https://github.com/blefnk/garden/discussions/6) if you notice any minor issues in the project, such as outdated information, broken links, grammatical or spelling mistakes etc. in the README.md or another files.

_You can also check [our another free Next.js 13 starter](https://github.com/blefnk/reliverse). This one starter, a monorepo, offers you tech used in the current starter plus also: Turborepo/Turbopack, Prisma, Valibot, Lucia, Clerk, and much more. It means there we're very experimentally trying to implement all the important and most used tech together._

## 03. We are Waiting for Your Help

The project has a huge plans! So we are very appreciate any help! Here is the things for the people, who are searching to make their own commits, like GitHub beginners.

**Currently we are very interested in the help with:**

1. README.md needs to be fully improved.
2. Some toaster notifications doesn't work.
3. Sometimes next-intl error is randomly shown in console.
4. Stripe is not fully finished and may not work in some places.

You can also use `Cmd/Ctrl+Shift+F` in VSCode to search for `todo:`.

Please **check [Commits](https://github.com/blefnk/garden/issues) tab** to see another things where help is needed.

Besides it, check the v2.0.0-v3.0.0 Project Roadmap below to get know what things can be implemented in this Next.js starter.

## 04. Project Features | Roadmap | Checklists

### 04.01. Roadmap Checklist For **v1.0.0** Release

This roadmap outlines the key features and improvements to be implemented in this Next.js starter. If mistakes here, please create an issue.

1. [x] 🌍 **[Next.js 13.5](https://nextjs.org)**, **[React 18](https://react.dev)**, and **[TypeScript 5](https://typescriptlang.org)** used as core.
2. [x] ✅ Perfectly configured `next.config.mjs` with i18n and mdx support implementation.
3. [x] 📚 Everything is tried to be as well-documented as possible, and beginner-friendly.
4. [x] 🧱 Nicely configured and explained `middleware.ts` for i18n and next-auth.
5. [x] 🍭 Cleanly composed modern user interface, built with **[Radix](https://radix-ui.com)** and stunning UI components, all thanks to **[shadcn/ui](https://ui.shadcn.com)** and **[@auth/drizzle-adapter](https://authjs.dev/reference/adapter/drizzle)**..
6. [x] 🌐 Comprehensive implementations for i18n, means internationalization, with 8 languages included, using **[next-intl](https://next-intl-docs.vercel.app/docs/getting-started/app-router-server-components)**, some things from **[next-international](https://github.com/QuiiBz/next-international)**, but in the future we want to use our very own solutions.
7. [x] 🦫 **[Drizzle ORM](https://orm.drizzle.team)** and **[DrizzleKit](https://orm.drizzle.team/kit-docs/overview)** configured for serverless-first environments, for services like **[Neon](https://neon.tech)** and **[Vercel Postgres](https://vercel.com/docs/storage/vercel-postgres)**.
8. [x] 🦺 Usage of [Cuid2](https://github.com/paralleldrive/cuid2) to automatically generate collision-free modern IDs in the database models.
9. [x] 📊 Data fetching, which works on server and client, using **[tRPC](https://trpc.io)** & **[TanStack Query](https://tanstack.com/query)**.
10. [x] ✳️ Awesomely crafted VSCode settings and recommended extensions.
11. [x] 🌿 Authentication via **[NextAuth](https://authjs.dev)** (previously known as NextAuth.js and next-auth) with account linking implementation.
12. [x] 🛍️ Implementation of storefront with products, categories, and subcategories.
13. [x] 🧫 Indicator implementation for TailwindCSS screen sizes.
14. [x] 💰 Implementation of user subscriptions and checkout system via **Stripe**.
15. [x] 💮 Optimized **[Next.js Metadata API](https://nextjs.org/docs/app/building-your-application/optimizing/metadata)** for SEO handling, with file-system handlers.
16. [x] 💻 Blog implementation by using **MDX** files.
17. [x] 📖 Well written `README.md`, and is beginner-friendly first, including descriptions for the **[environment variables](https://nextjs.org/docs/basic-features/environment-variables)**.
18. [x] 🤩 Using **[TailwindCSS](https://tailwindcss.com)** for utility-first CSS.
19. [x] 🦦 Using **[EsLint](https://eslint.org)** with **[Prettier](https://prettier.io)** for readable, safe code.
20. [x] 🅰️ The beautiful implementation of font system, using **[Inter](https://rsms.me/inter)** and other typefaces.
21. [x] 🔀 Type-safety validations for project schemas and UI fields via **[Zod](https://zod.dev)**.

### 04.02. Roadmap Checklist For **v2.0.0** Release

This roadmap outlines the key features and improvements to be implemented in this Next.js starter. Not marked things may be already configured, but was not tested too much. If mistakes here, please create an issue.

1. [ ] 🦾 Partially complete BA11YC (Bleverse Accessability Convention) checklist and garden's own checklist section (find it below for `1.0.0`-`2.0.0`).
2. [ ] 🧁 Full [Bun](https://bun.sh) support and compatibility.
3. [ ] 💎 Advanced implementation of user subscriptions and checkout system via **Stripe**.
4. [ ] 🏬 Advanced implementation of storefront with products, categories, and subcategories.
5. [ ] 🤖 Implement AI features and chat, using e.g [Vercel AI SDK](https://sdk.vercel.ai/docs), to this starter (read: [Introducing the Vercel AI SDK](https://vercel.com/blog/introducing-the-vercel-ai-sdk)).
6. [ ] 🌩️ Full Next.js Edge things support and compatibility.
7. [ ] 📝 Comment system for products, in types of Review and Question.
8. [ ] 🔔 Advanced notification system, like toaster, pop up, and page.
9. [ ] 💠 Advanced 404 Not Found page with full internationalization support.
10. [ ] 🪪 Advanced sign-up/sign-in/restoring by using email-password and magic-links.
11. [ ] 🏎️ Try to optimize everything in the app to make also faster dev cold start and build.
12. [ ] 🪄 Use [Kysely](https://kysely.dev/) with Drizzle to get the full type-safety for TypeScript SQL query builder.
13. [ ] 🔀 Implement some useful tests, including if possible stress-tests to simulate and check app performance, when many people uses the app at one time.
14. [ ] 🔗 Minimal how it's possible HTML tags nesting. And also correct ways of using HTML tags.
15. [ ] ⚙️ Beautifully configured `app.ts`, a single config to replace them all.
16. [ ] ✒️ Gorgeously written comments everywhere, keeping the code still clean.
17. [ ] 🐎 Nicely configured `package.json` with correctly installed and sorted packages in `dependencies` and `devDependencies`.
18. [ ] ✨ Admin dashboard with stores, products, orders, subscriptions, and payments.
19. [ ] ♨️ Advanced indicators implementations for installed packages, environment variables, and improvements to TailwindCSS screen sizes.
20. [ ] 🦄 Project is clean from duplicates, like files, components, etc.
21. [ ] 🌬️ Advanced theme switching without flashing, using Tailwind Dark Mode with [React Server Side support](https://michaelangelo.io/blog/darkmode-rsc) and dynamic cookies.
22. [ ] ⬆️ File uploads implementation using [UploadThing](https://uploadthing.com) and [Cloudinary](https://cloudinary.com).
23. [ ] 🏟️ Everything in the project is sorted in ascending order, unless a different sort is required somewhere.
24. [ ] 🌀 Comprehensive configured Next.js 13 App Router, API routes handled by Route Handlers, with the RSCs and everything all new.
25. [ ] 💫 Use where it make sense the [next-usequerystate](https://github.com/47ng/next-usequerystate) library ([read the article](https://francoisbest.com/posts/2023/storing-react-state-in-the-url-with-nextjs)).
26. [ ] 🌅 Using every usable [TanStack](https://tanstack.com) libraries.
27. [ ] 🌌 Even more comprehensive i18n implementation, with using country and locale codes, supporting 20 languages. Every language must be verified by native speakers after machine translation.
28. [ ] 🌞 Clean from any of unused things, like packages, libs, vars, etc.
29. [ ] 🍪 Minimal possible number of cookies, to be ready for cookie-free future, with cookie notification implementation.
30. [ ] ⭐ Complete advanced sign up and sign in pages both with socials and classic form methods.
31. [ ] 🐧 The most beginner-friendly and prettyful starter, how it's possible.
32. [ ] 🦊 Every possible page is wrapped using predefined shell wrappers.
33. [ ] ❄️ Project is clean from `@ts-expect-error` or similar things.
34. [ ] 🐭 Minimal number of config files, how it possible.
35. [ ] 🌳 Using `function` instead `const` for components everywhere, to make it more cleaner and beginners-friendly.
36. [ ] 🦑 Absolute paths are used wherever it makes sense.
37. [ ] 🐳 [Jest](https://jestjs.io) testing, optimized for Next.js 13.
38. [ ] 📰 Email verification and newsletter subscription and email marketing via [Resend](https://resend.com) and [React Email](https://react.email).
39. [ ] 🤖 Ultimate type-safety using strict-mode in [TypeScript](https://typescriptlang.org), typedRoutes, zod, middlewares etc.
40. [ ] 🪟 Pop-ups like: cookies/GDPR notification (with corresponding management settings page), Google floating notification about fast login, etc.
41. [ ] 🤝 Seller and customer workflows.
42. [ ] 🦾 Accessability everywhere, for app users UI (User Interface) with UX (User Experience), and for developers DX (Developer Experience). And with usability, but still with beauty in the mind.
43. [ ] 🌉 Better `middleware.ts` configuration with multi-mw implementation.
44. [ ] 🪶 All website languages must be free of grammatical and spelling errors and written in accordance with the newest rules of each language.
45. [ ] 😜 Minimal possible references to garden, Bleverse, etc. to be really easy-starter. Using variables from `app.ts` everywhere.
46. [ ] 🎯 Maximum possible for the app performance points on services like Google PageSpeed Insights. Also app must pass all hard tests.

### 04.03. Roadmap Checklist For **v3.0.0** Release

This roadmap outlines the key features and improvements to be implemented in this Next.js starter. Not marked things may be already configured, but was not tested too much. If mistakes here, please create an issue.

1. [ ] 🦿 Fully completed BA11YC (Bleverse Accessability Convention) checklist.
2. [ ] 🔥 Implement even more feature to be something more than just a e-commerce store to become a starter for all kind of websites.
3. [ ] 🚀 Implement all the useful things from [Next.js Weekly](https://nextjsweekly.com/issues) into this starter.
4. [ ] 💚 Discuss, document, and use the conventions about e.g. single naming case style for file and variables.
5. [ ] ❇️ Achieve as few files as possible by combining similar things, etc.
6. [ ] 🌐 Translate README.md and/or similar files into other languages.

### 04.04. garden `1.0.0`-`2.0.0` Accessability Checklist

_The section will be expanded soon._

1. [ ] 😇 Maximum accessability **for users**. Accessability **for developers**, both beginners and seniors. Accessability **for bots**, like `[Googlebot](https://developers.google.com/search/docs/crawling-indexing/googlebot)` or `[PageSpeed Insights](https://pagespeed.web.dev/) Crawler`. Accessability **for everyone**.
2. [ ] 😺 Using some interesting recommendations from \*\*[Material Design 3](https://m3.material.io) and other design systems.

## 06. Table of Contents

<!-- TOC start -->

- [garden Next.js 13.5: Store | Landing | Dashboard (Template 2023)](#garden-nextjs-135-store--landing--dashboard-template-2023)
  - [01. Follow Us Everywhere](#01-follow-us-everywhere)
  - [02. The Bleverse Framework for the Web](#02-the-bleverse-framework-for-the-web)
  - [03. We are Waiting for Your Help](#03-we-are-waiting-for-your-help)
  - [04. Project Features | Roadmap | Checklists](#04-project-features--roadmap--checklists)
    - [04.01. Roadmap Checklist For **v1.0.0** Release](#0401-roadmap-checklist-for-v100-release)
    - [04.02. Roadmap Checklist For **v2.0.0** Release](#0402-roadmap-checklist-for-v200-release)
    - [04.03. Roadmap Checklist For **v3.0.0** Release](#0403-roadmap-checklist-for-v300-release)
    - [04.04. garden `1.0.0`-`2.0.0` Accessability Checklist](#0404-garden-100-200-accessability-checklist)
  - [06. Table of Contents](#06-table-of-contents)
  - [07. Recommendations \& Why We Use This And That](#07-recommendations--why-we-use-this-and-that)
  - [08. Bun Support \& Compatibility](#08-bun-support--compatibility)
  - [09. Internationalization (i18n)](#09-internationalization-i18n)
  - [10. Project Configuration](#10-project-configuration)
  - [11. NextAuth.js Authentication](#11-nextauthjs-authentication)
  - [12. Environment variables](#12-environment-variables)
  - [13. Getting Started Locally](#13-getting-started-locally)
  - [14. Database](#14-database)
    - [1. `pnpm db` automatization](#1-pnpm-db-automatization)
    - [2. Making changes to the database schema (learn more)](#2-making-changes-to-the-database-schema-learn-more)
    - [tRPC](#trpc)
  - [15. How to Deploy the Project](#15-how-to-deploy-the-project)
  - [16. Migration From Similar Starter](#16-migration-from-similar-starter)
  - [17. UI components](#17-ui-components)
  - [18. Linting / Checking the codebase](#18-linting--checking-the-codebase)
    - [18.01. Linting](#1801-linting)
    - [18.02. Type Checking](#1802-type-checking)
    - [18.03. Formatting with Prettier](#1803-formatting-with-prettier)
    - [18.04. Testing](#1804-testing)
  - [19. Say Thank You to Them](#19-say-thank-you-to-them)
  - [19. Recommended Things to Learn](#19-recommended-things-to-learn)
  - [20. Contributing and Credits](#20-contributing-and-credits)
    - [20.01. Code Guide and Recommendations](#2001-code-guide-and-recommendations)
  - [21. Interesting Things](#21-interesting-things)
    - [21.01. Typical Workflow How App Works](#2101-typical-workflow-how-app-works)
    - [21.02. What garden means?](#2102-what-garden-means)
  - [22. License](#22-license)

<!-- TOC end -->

_The section will be improved soon. You can use [this service](https://derlin.github.io/bitdowntoc) (with GitHub preset selected) if you need to update current `README.md` file. VS Code extensions also has good automatic TOC generating._

## 07. Recommendations & Why We Use This And That

_The section will be implemented soon._

Currently we don't use Contentlayer. It still unstable with Windows. For this reason we currently trying to implement options for choosing to use or not to use Contentlayer in the `src/app.ts` configuration file. In the future, maybe, we are develop our very own solution for the content writing. External providers, like Sanity, also possibly will be implemented in the future with corresponding options to enable/disable.

We recommend to clean browser cache and delete `.next` folder from time to time, to ensure that if everything works as expected.

![Bleverse garden OG Image](/public/og-image.png)

## 08. [Bun](https://bun.sh) Support & Compatibility

The `garden` already can be used with some `Bun` awesome features. Currently we recommend to use `pnpm` for this starter. We try to ship the full as possible Bun support and compatibility as soon as the Windows binaries will be released.

_The section will be expanded soon._

## 09. Internationalization (i18n)

At Bleverse, the approach to multilingualism is a separate Holy Grail. We love to talk about this topic. In the future, we will write a lot of interesting things about Next.js 13 App Router internationalization.

**Hint:** Use `pnpm lint:i18n` to check if your i18n files are okay. It also tries to fix some things if possible with tools e.g. it provides the ascending sort. If you got nothing on output it means everything is okay.

Currently, all languages have machine translation. In the future, we plan to have them checked by native speakers.

Please note that the project currently has i18n messages from another of our open source projects. They will be removed in the next few days.

This section will be expanded accordingly in the future. Stay tuned!

**Current Locales** _(asc sort)_

- 🇩🇪: "German",
- 🇬🇧: "English",
- 🇪🇸: "Spanish",
- 🇫🇷: "French",
- 🇮🇹: "Italian",
- 🇵🇱: "Polish",
- 🇹🇷: "Turkish",
- 🇺🇦: "Ukrainian",

## 10. Project Configuration

The `src/app.ts` file contains essential configuration used to modify the contents and settings of the website.

- Manage the content displayed on the website.
- Customize various settings, such as disabling the theme toggle.
- Manage general site-wide information.

## 11. NextAuth.js Authentication

Setting up auth is easy as a breeze...

_The section will be implemented soon._

## 12. Environment variables

**This section will be more clean and improved soon. Please stay tuned.**

Define default values for environment variables required to run the app. Never store secrets in `.env.example` file. Store sensitive information in a `.env` file only. Do not commit these local env files to source control. If you're new or cloning the repo, use the `.env.example` file as a template to create your `.env` file. Keep this file current when adding new variables to `.env`. Use Caution When Committing. Remember that this file will be committed to version control. Ensure it does not contain secrets. When cloning this repo, create a copy named `.env` and add your secrets. When adding new environment variables, update the schema in `/src/env.mjs` accordingly.

You can [read more about environment variables here](https://nextjs.org/docs/app/building-your-application/configuring/environment-variables).

`.env.example` file was simplified so much with the 1.0.0 garden release, and will be even more simplified in the next versions. We plan to implement if the specific value is not specified, then the app will still be fine compiled by disabling this env related functional.

**🟢 Explanations to [`.env.example`](https://github.com/blefnk/garden/blob/main/.env.example) file (COPY DATA FROM THERE TO NEW `.env` FILE)**

1. **Database**

   You can get database credentials from [Neon](https://neon.tech) or [Vercel Postgres](https://vercel.com/docs/storage/vercel-postgres). Another PostgreSQL serverless supported providers also may work.
   `DATABASE_URL="postgresql://<...>"`

2. **[Github](https://github.com/settings/applications/new)**

   - `GITHUB_CLIENT_ID="fake"`
   - `GITHUB_CLIENT_SECRET="fake"`

3. **[Discord](https://discord.com/developers/applications)**

   - `DISCORD_CLIENT_ID="fake"`
   - `DISCORD_CLIENT_SECRET="fake"`

4. **[Google](https://console.cloud.google.com/apis/credentials)**

   - `GOOGLE_CLIENT_ID="fake"`
   - `GOOGLE_CLIENT_SECRET="fake"`

5. **[Email](https://postmarkapp.com)**

   - `SMTP_FROM="fake"`
   - `POSTMARK_API_TOKEN="fake"`
   - `POSTMARK_SIGN_IN_TEMPLATE="fake"`
   - `POSTMARK_ACTIVATION_TEMPLATE="fake"`

6. **JWTs ([@docs](https://next-auth.js.org/configuration/options#nextauth_secret)) and default site url (use domain in production)**

   You can generate Auth.js JWTs secret using [ConvertSimple.com](https://convertsimple.com/random-base64-generator) website.

   - `NEXTAUTH_SECRET="ChangeMeToSomethingRandom"`
   - `NEXT_PUBLIC_APP_URL="http://localhost:3000"`

7. **[RESEND](https://resend.com)**

   Resend API Key found at [@link](https://resend.com/api-keys)
   We need to register a domain with Resend to send emails from
   Register a domain at [@link](https://resend.com/domains)
   _Or we can use this email provided by resend for only testing:_

   - `EMAIL_FROM="Resend <onboarding@resend.dev>"`
   - `RESEND_API_KEY="re_"`

8. **[STRIPE CREDENTIALS](https://stripe.com)**

   These values should be replaced with production keys on the deployed site to accept real payments.
   _Stripe Publishable Key and Secret Key found at [@link](https://dashboard.stripe.com/test/apikeys)_
   `NEXT_PUBLIC_STRIPE_PUBLISHABLE_KEY="pk_test_"`
   `STRIPE_API_KEY="sk_test_"`
   Stripe Webhook Secret found at [@link](https://dashboard.stripe.com/test/webhooks/create?endpoint_location=local).
   _The webhook URL should be: https://<INSERT_YOUR_DOMAIN_HERE>/api/stripe/webhook_
   `STRIPE_WEBHOOK_SECRET="whsec_"`
   _Stripe Product and Price IDs for your created products_
   _found at [@link](https://dashboard.stripe.com/test/products)_
   `STRIPE_ADVANCED_MONTHLY_PRICE_ID="price_"`
   _Subscription Pricing_
   _Stripe Product and Price IDs for your created products_
   _found at [@link](https://dashboard.stripe.com/test/products)_
   `STRIPE_BASIC_MONTHLY_PRICE_ID="price_"`

9. **OTHER**

   **[FILE UPLOADING](https://uploadthing.com)**

   - `UPLOADTHING_SECRET="sk_live_"`
   - `UPLOADTHING_APP_ID="fake"`

   **[LOGLIB](https://loglib.io)**

   - `LOGLIB_API_KEY="site_"`
   - `LOGLIB_SITE_ID="_"`

## 13. Getting Started Locally

Install Git.

```bash
https://git-scm.com
```

Install Node.js.

_Windows:_

```bash
https://github.com/coreybutler/nvm-windows/releases/download/latest/nvm-setup.exe
```

_Linux:_

```bash
https://github.com/nvm-sh/nvm#installing-and-updating
```

Install PNPM.

```bash
corepack enable
```

Clone the repository.

```bash
git clone https://github.com/blefnk/garden.git
```

Get your environment variables in check and update the variables.

```sh
cp .env.example .env
```

Get all dependencies sorted.

```sh
pnpm install
```

Bring your database to life with pushing the database schema.

```bash
pnpm db:generate
pnpm db:push
```

Let the server magic begin! The app will be running at <http://localhost:3000>.

```sh
pnpm dev
```

Finally, start the Stripe webhook listener for Stripe to work.

```bash
pnpm stripe:listen
```

You can lint the project using `pnpm lint`.

You can build also using `pnpm build` or `pnpm build:tsc`,
`pnpm build:tsc` shows you all typescript errors if exist.

_And, ladies and gentlemen, it's perfectly ready!_

**Hint:** You can update all project packages to theirs latest stable versions by using `pnpm latest` command.

## 14. Database

Drizzle ORM in the garden is set up to use be ready for serverless environments by default. The project build with [PostgresJS](https://orm.drizzle.team/docs/quick-postgresql/postgresjs) and uses `[pg](https://orm.drizzle.team/docs/quick-postgresql/node-postgres)` ([docs](https://neon.tech/docs/serverless/serverless-driver#pool-or-client)), so any PostgreSQL-like supported database services will work. Everything in the project currently set up serverless using [@neondatabase/serverless](https://orm.drizzle.team/docs/quick-postgresql/neon). Simply set `DATABASE_URL` in created by you `.env` file and you are ready to go. You can learn more about it in the [env related section](#12-environment-variables) of this readme.

If you want to use another serverless provider, like [Vercel Postgres](https://vercel.com/docs/storage/vercel-postgres), just instead of [@neondatabase/serverless](https://orm.drizzle.team/docs/quick-postgresql/neon) package use [@vercel/postgres](https://orm.drizzle.team/docs/quick-postgresql/vercel) everywhere and you can still use `[postgres](https://orm.drizzle.team/docs/quick-postgresql/postgresjs)` or `[pg](https://orm.drizzle.team/docs/quick-postgresql/node-postgres)` driver in the`drizzle.config.ts` file.

If you need a serverful architecture (which [is not recommended](https://www.prisma.io/dataguide/serverless/traditional-vs-serverless-databases) and it is [the old thing](https://www.locofy.ai/blog/serverful-vs-serverless)), just remove ``from the app and just use only [PostgresJS](https://orm.drizzle.team/docs/quick-postgresql/postgresjs), and set the driver to `[pg](https://orm.drizzle.team/docs/quick-postgresql/node-postgres)` in the`drizzle.config.ts` file.

### 1. `pnpm db` automatization

This project exposes a package.json script for accessing drizzle-kit via `pnpm db:<command>`. This script handles all environment variable mapping automatically via `dotenv-cli`, so you don't have to think about it. You should always try to use this script when interacting with drizzle-kit locally.

**Note:** Don't delete any files in `migrations` folder manually, please refer to [pnpm db:drop](https://orm.drizzle.team/kit-docs/commands#drop-migration) command.

### 2. Making changes to the database schema ([learn more](https://orm.drizzle.team/docs/sql-schema-declaration))

1. [optional] Make changes to your database by declaring or modifying the `*.ts` files of prototypes in `src/data/db` folder.

2. [optional] After declaring or modifying database prototypes you will then use `pnpm db:generate`.

3. [recommended] You can run the `pnpm db:push` after a double checking the new generated files in the `drizzle` folder. If something wrong **please don't delete the files inside `drizzle` folder by yourself and use `pnpm db:drop` instead, manual removing can crash Drizzle, [as its developers said](https://orm.drizzle.team/kit-docs/conf#migrations-folder)**. And when you feel comfortable with the changes, you can then send the SQL generated files code to your database provider. _Recommended to run this after cloning the repo and creating your new database in the chosen by you database service._


changelog.md
# Project Changelog

**This file will be updated in the future.**

## [@garden/commits](https://github.com/blefnk/garden/commits/main)

## [0.1.0](https://github.com/blefnk/garden/releases/tag/0.0.1)

- [🌱 feat(app): initial deploy](https://github.com/blefnk/garden/commit/4099c7c7e444b3e5782487355a508fdfdb3a14cd)


src/app.ts
/**
 * To reduce the number of config files, we aim to combine everything into a single file.
 * Materials about @satisfies: https://youtu.be/49gHWuepxxE, https://youtu.be/G1RtAmI0-vc
 */

import { MainMenuItem, type FooterItem } from "~/types";

import { ContentSection, HeroHeader } from "~/server/config/appts";
import { productCategories } from "~/server/config/products";
import { networks } from "~/server/config/socials";
import { slugify } from "~/server/utils";

export const appts = {
  name: "garden",
  social: networks({
    youtube: "@bleverse_com",
    discord: "Pb8uKbwpsJ",
    facebook: "groups/bleverse",
    twitter: "blefnk",
    github: "blefnk",
  }),
  // currently not recommended to use debug
  debug: false,
};

export default appts;

export type articles = {
  type?: "blog_news" | "blog_only" | "news_only" | undefined;
  articles_provider?: "default" | "contentlayer" | undefined;
};

export type SiteConfig = typeof siteConfig;

const links = {
  twitter: "https://x.com/blefnk",
  github: "https://github.com/blefnk/garden",
  githubAccount: "https://github.com/blefnk",
  discord: "https://discord.gg/Pb8uKbwpsJ",
  facebook: "https://facebook.com/groups/bleverse",
};

export const contactConfig = {
  email: "blefnk@gmail.com",
};

export const REPOSITORY_OWNER = "blefnk";
export const REPOSITORY_NAME = "garden";
export const REPOSITORY_URL = `https://github.com/${REPOSITORY_OWNER}/${REPOSITORY_NAME}`;
export const baseUrl = "https://garden.bleverse.com";

export const BASE_URL =
  process.env.NODE_ENV === "production" ? baseUrl : "http://localhost:3000";
export const BRAND_NAME = "garden";
export const BRAND_DESCRIPTION =
  "Next.js 13 free store and dashboard template. It helps you build great eCommerce and SaaS apps faster than ever. Get it now!";

export const OWNER_ROLE = "owner";
export const ADMIN_ROLE = "admin";
export const MEMBER_ROLE = "member";

export const TRIAL_LENGTH_IN_DAYS = 7;
export const ROLES = [OWNER_ROLE, ADMIN_ROLE, MEMBER_ROLE] as const;

export const settings = {
  internationalizationEnabled: true,
  themeToggleEnabled: true,
};

export const siteConfig = {
  name: "garden",
  shortName: "garden",
  author: "Bleverse",
  description:
    "NextJS 13 free starter: store, landing, dashboard. It helps you build great eCommerce and SaaS apps faster than ever. Get it!",
  company: {
    name: "Bleverse",
    link: "https://bleverse.com",
    email: "blefnk@gmail.com",
    twitter: "@blefnk",
  },
  handles: {
    twitter: "@blefnk",
  },
  keywords: [
    "App Router",
    "Blefonix",
    "Bleverse",
    "Drizzle Orm",
    "Landing Page",
    "Lemon Squeezy",
    "Next.js 13.5",
    "Nextjs",
    "Open Source",
    "Parallel Routes",
    "PostgreSQL",
    "Radix Ui",
    "React",
    "garden",
    "Server Actions",
    "Server Components",
    "Shadcn/UI",
    "Starter",
    "Stripe",
    "T3 Stack",
    "Tailwind Css",
    "Template",
    "Tools",
    "Utils",
  ],
  url: {
    base: baseUrl,
    author: REPOSITORY_OWNER,
  },
  ogImage: `${baseUrl}/og-image.png`,
  mainNav: [
    {
      title: "Lobby",
      items: [
        {
          title: "Products",
          href: "/products",
          description: "All the products we have to offer.",
          items: [],
        },
        {
          title: "Build a Look",
          href: "/custom/clothing",
          description: "Build your own custom clothes.",
          items: [],
        },
        {
          title: "Blog",
          href: "/blog",
          description: "Read our latest blog posts.",
          items: [],
        },
      ],
    },
  ] satisfies MainMenuItem[],
  links,
  footerNav: [
    {
      title: "Bleverse",
      items: [
        {
          title: "Community",
          href: "https://bleverse.com",
          external: true,
        },
        {
          title: "MF Piano",
          href: "https://mfpiano.org",
          external: true,
        },
        {
          title: "Peresfer",
          href: "https://peresfer.com",
          external: true,
        },
        {
          title: "garden",
          href: "https://garden.bleverse.com",
          external: true,
        },
      ],
    },
    {
      title: "Help",
      items: [
        {
          title: "Contact",
          href: "/contact",
          external: false,
        },
        {
          title: "Privacy",
          href: "/privacy",
          external: false,
        },
        {
          title: "Terms",
          href: "/terms",
          external: false,
        },
        {
          title: "About",
          href: "/about",
          external: false,
        },
      ],
    },
    {
      title: "Social",
      items: [
        {
          title: "Facebook",
          href: links.facebook,
          external: true,
        },
        {
          title: "Discord",
          href: links.discord,
          external: true,
        },
        {
          title: "Twitter",
          href: links.twitter,
          external: true,
        },
        {
          title: "Github",
          href: links.githubAccount,
          external: true,
        },
      ],
    },
    {
      title: "Github",
      items: [
        {
          title: "Nomaders",
          href: "https://github.com/blefnk/nomaders",
          external: true,
        },
        {
          title: "Reliverse",
          href: "https://github.com/blefnk/reliverse",
          external: true,
        },
        {
          title: "garden",
          href: "https://github.com/blefnk/garden",
          external: true,
        },
        {
          title: "Utilities",
          href: "https://github.com/blefnk/utils",
          external: true,
        },
      ],
    },
  ] satisfies FooterItem[],
};

export const heroHeader: HeroHeader = {
  header1: `Next.js 13 Store & Dashboard Template`,
  header2: `Build Great eCommerce and SaaS Faster`,
  subheader: `shadcn/ui, Link, App Router, TypeScript, T3, Stripe, NextAuth.js, Tailwind,
  Drizzle, Zod, RSC, SWC, tRPC, NextAuth, Server Actions, Lucide Icons,
  etc.`,
};

export const featureCards: ContentSection = {
  header: `Powered by`,
  subheader: `What Makes garden Possible`,
  content: [
    {
      text: `Next.js`,
      subtext: `The React Framework`,
    },
    {
      text: `shadcn/ui`,
      subtext: `Beautifully Designed Components`,
    },
    {
      text: `Vercel`,
      subtext: `Develop. Preview. Ship.`,
    },
  ],
};

export const features: ContentSection = {
  header: `Features`,
  subheader: `Why You Need to Download garden`,
  content: [
    {
      text: `SEO Optimized`,
      subtext: `Improved website visibility on search engines`,
    },
    {
      text: `Highly Performant`,
      subtext: `Fast loading times and smooth performance`,
    },
    {
      text: `Easy Customization`,
      subtext: `Change your content and layout with little effort`,
    },
  ],
};
